posts/2019-01-22-the-5-raisons-detre-of-testing.md

```diff
@@ -0,0 +1,219 @@
+```
+
+The usage is _straightforward_:
+
+* if `nameOfTheBreed` is `persian` then you read `cat.png` and return the
+  resulting bytes and `200 OK`.
+* if `nameOfTheBreed` is `siamese` then you read `CAT.png` and return the
+  resulting bytes and `200 OK`.
+* if `nameOfTheBreed` is anything else, you return zero bytes and `200 OK`
+  (company policy GTE-421-f, introduced in 2003, forbids you from using
+  `404`s).
+
+</div>
+
+Because we're dealing with software, anything that can go wrong will go wrong.
+Let's see how tests can help you damage control. Below are five reasons why I
+believe tests are important, in no particular order.
+
+## Ensure that the code you've written works
+
+This may be the most straightforward reason to write tests. However, while it's
+valuable to know that whatever you've just written works, it's probably the
+least important reason.
+
+> Testing that code works is the _raison d'être_ of tests!!! How dare you say
+> that it's not important???
+>
+> -- Jimmy from Nebraska
+
+Don't get it twisted Jimmy, I didn't say it was unimportant: the other reasons
+are just _more_ important. Most of the time you'll eventually try the feature
+(open the website, click `Siamese`, watch in awe as your cat appears on the
+screen) after having written the code -- which makes tests redundant. You will
+catch bugs like "trying to open `cat.jpg` instead of `cat.png`" straight away
+anyway.
+
+Running tests as you write the code is mostly a time saver when you develop a
+feature, as you (usually) have a single command to run in order to check that
+you've done things correctly. Also, by writing tests close to your spec, it
+acts as a deadline (once your tests are green, you're done, you can move on to
+something else). Moreover by _first_ writing a (failing) test when tasked with
+fixing a bug you ensure that the bug actually exists. This is typical
+[TDD](https://en.wikipedia.org/wiki/Test-driven_development).
+
+## Ensure that the code you've written works elsewhere
+
+
+<div class="story">
+You wrote your feature and Tommy wrote his part. Feature "ninja" got merged
+into the master branch, after you diligently tested it out by hand. You go home
+that evening, and as you're about to start watching the third season of The
+Wire (for the second time, because you first assumed you didn't need English
+subtitles) you get an angry phone call from Jason, the sales rep working late
+nights because he's demoing the product to people in another timezone. Jason
+explains that he can only get a picture of the Persian cat. Even when he picks
+the Siamese option from the dropdown. The debugging process goes like this:
+
+> You: What OS are you using?
+>
+> Jason: What's an _awe-hess_?
+
+At which point you realize that Jason is running your code on his Windows
+machine. Which, unfortunately enough, cannot tell the difference between
+`cat.png` and `CAT.png`. You sigh, put the remote down and go back to the
+drawing board.
```

Comment:
I knew it! You little bugger. And it was so easy for you to diagnose the problem because you knew it too!

Reply:
Maybe you would add a Windows machine to the CI and discover it that way?